torch/distributed/rpc/api.py

```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
     return states.gathered_objects
 
 
+# detect if any worker has invalid map_location configurations, and return
+# names of failed workers
+@_require_initialized
+def _setup_map_locations(map_locations):
+    agent = _get_current_rpc_agent()
+    if not map_locations:
+        return
+    elif not isinstance(agent, TensorPipeAgent):
+        raise ValueError(
+            "map_locations attribute only works with TensorPipe RPC Backend, "
+            f"but the current agent type is {type(agent)}"
+        )
+
+    def check_one_worker(name, map_locations, all_device_counts):
+        device_count = all_device_counts[name]
+        wrong_worker_names = set(map_locations) - set(all_device_counts)
+        if wrong_worker_names:
+            raise ValueError(f"Wrong worker names: {wrong_worker_names}")
+        for worker_name in all_device_counts:
+            remote_device_count = all_device_counts[worker_name]
+            if worker_name in map_locations:
+                map_location = map_locations[worker_name]
+                key_set = set(map_location.keys())
+                val_set = set(map_location.values())
+                if not all([
+                    len(map_location) == len(key_set),
+                    len(map_location) == len(val_set),  # check 1-to-1 mapping
+                    min(key_set) >= -1,
```

Comment:
Yep, we can remove that in the first version, but this might be a useful feature in just the driver/master does not have GPU, or if the model cannot fit into its GPU.

Reply:
Yes, as soon as TensorPipe supports D2H and H2D transfers we should start allowing CPUs in the device map, but until then the only way to do such a transfer is to add an intermediate hop on the CPU (i.e., leader copies across network from its CPU to worker's CPU, and then workers copies locally from CPU to GPU) and this is something that users can already do manually if they so wish.